server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -380,12 +381,27 @@ static double getMaxIndexLoadWithinCoolingPeriod(
         // assume the current write index load is the highest observed and look back to find the actual maximum
         double maxIndexLoadWithinCoolingPeriod = writeIndexLoad;
         for (IndexWriteLoad writeLoad : writeLoadsWithinCoolingPeriod) {
-            double totalIndexLoad = 0;
+            // the IndexWriteLoad stores _for each shard_ a shard average write load ( calculated using : shard indexing time / shard
+            // uptime ) and its corresponding shard uptime
+            //
+            // to reconstruct the average _index_ write load we recalculate the shard indexing time by multiplying the shard write load
+            // to its uptime, and then, having the indexing time and uptime for each shard we calculate the average _index_ write load using
+            // (indexingTime_shard0 + indexingTime_shard1) / (uptime_shard0 + uptime_shard1)
+            // as {@link org.elasticsearch.index.shard.IndexingStats#add} does
+            double totalShardIndexingTime = 0;
+            long totalShardUptime = 0;
             for (int shardId = 0; shardId < writeLoad.numberOfShards(); shardId++) {
                 final OptionalDouble writeLoadForShard = writeLoad.getWriteLoadForShard(shardId);
-                totalIndexLoad += writeLoadForShard.orElse(0);
+                final OptionalLong uptimeInMillisForShard = writeLoad.getUptimeInMillisForShard(shardId);
+                if (writeLoadForShard.isPresent()) {
+                    assert uptimeInMillisForShard.isPresent();
+                    double shardIndexingTime = writeLoadForShard.getAsDouble() * uptimeInMillisForShard.getAsLong();
+                    long shardUptimeInMillis = uptimeInMillisForShard.getAsLong();
+                    totalShardIndexingTime += shardIndexingTime * shardUptimeInMillis;
+                    totalShardUptime += shardUptimeInMillis;
+                }
```

Comment:
Was this change for the average correction something that came up elsewhere? Was there a bug where we were calculating it incorrectly before this?

Reply:
The problem is that my math was poor before and I hadn't realised that when computing the **index** load (as returned by the indexing stats) we're doing an average per index: i.e. we sum the indexing time for all the shards and divide this **sum** by the sum of all shards uptime ( https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/shard/IndexingStats.java#L109 )

However, the values we store in the index metadata is the write load / shard (that is calculated by dividing the indexing time to the uptime) and it's corresponding uptime. 

So the math we had before here (when we just summed up the write load / shard) was not correct as that didn't represent the **index** average load. 

The current logic here reconstructs the logic we use in the indexing stats API ( https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/shard/IndexingStats.java#L109 ) to recompute the **index** load (i.e. we sum the indexing time for all the shards and divide this by the sum of all shards uptime)

I've added another unit tests that has various uptime and shard write load values https://github.com/elastic/elasticsearch/pull/106076/files#diff-64b9199161b79c41c9141c2b4c9cb5a55d24ea6d401e938262112927d432fa14R652